### PR TITLE
add iroh service

### PR DESCRIPTION
**Pull request by crimeminister**
## Description of changes

This is a service module for [iroh](https://iroh.computer), a content-addressable network filesystem similar to [IPFS](https://ipfs.tech) but aimed at greater scalability and performance. This module enables users to run an iroh node, enabling the distribution of data added locally as well as facilitating the dissemination of data stored in remote nodes participating in the global content distribution network. The service supports the distribution of data blobs and structured "documents", and features a cryptographic author and permission model.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by crimeminister** [comment]
@cameronfyfe I am slowly chipping away at an `iroh` service; I think I recall mentioning this plan a long while back and you expressing interest (might be a hallucination, though, in which case sorry for bugging you). If that's right, feel free to let me know what needs improving and/or fixing as I am new to module development.

I'm currently taking aim at `iroh 0.19.0` and trying to get it running with a systemd "dynamic user". The single current test is currently b0rked:

```
$ nix-build -A nixosTests.iroh
```
gives:
```
error: builder for '/nix/store/0vajl5gn0lhk0sgq1rpp1h71kr77wdlz-vm-test-run-iroh.drv' failed with exit code 1;
       last 10 log lines:
       > write # Caused by:
       > write #     iroh is not running, please start it
       > write # [    6.060457] su[899]: pam_unix(su:session): session closed for user alice
       > write: output:
       > Test "Add and retrieve data" failed with error: "command `su -- alice -l -c 'iroh blob add test-data'` failed (exit code 1)"
```
It's a work in progress...

**Comment by aanderse** [comment]
hello 👋 

thanks for contributing your time and energy to make NixOS better! 

could you please fill out the "motivation" section of the PR template? i ask because even after reading upstream documentation i am still not sure what the motivation for adding this as a service to NixOS is 🤷‍♂️ is the idea that this would be useful for developers working on a project? or useful to other software which might depend on `iroh`? maybe `iroh` is somehow directly useful to end users? etc...

i would love to learn more about what this software is useful for and how it might be useful in NixOS

thanks!

**Comment by crimeminister** [comment]
> hello 👋
> 
> thanks for contributing your time and energy to make NixOS better!
> 
> could you please fill out the "motivation" section of the PR template? i ask because even after reading upstream documentation i am still not sure what the motivation for adding this as a service to NixOS is 🤷‍♂️ is the idea that this would be useful for developers working on a project? or useful to other software which might depend on `iroh`? maybe `iroh` is somehow directly useful to end users? etc...
> 
> i would love to learn more about what this software is useful for and how it might be useful in NixOS
> 
> thanks!

Hi @aanderse, I added a little blurb to the PR description, let me know if it should be further fleshed out. The tl;dr is "like IPFS, but in Rust and maybe faster/more scalable".

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -4120,6 +4120,13 @@
     githubId = 1707779;
     name = "Chris Ertel";
   };
+  crimeminister = {
```
Comment (aanderse) [inline maintainers/maintainer-list.nix:4123]: this should be in a separate commit

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 2]
```diff
@@ -0,0 +1,275 @@
+{ config, lib, pkgs, utils, ... }:
+with lib;
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:2]: many people are highly discouraging this these days

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        default = null;
+      };
+
+      extraFlags = mkOption {
+        type = types.listOf types.str;
+        description = "Extra flags passed to the Iroh daemon";
+        default = [ ];
+      };
+
+      fileLogs = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:159]: this isn't referenced anywhere

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+  # The total set of relays is the merge of relayNodes (which are the default nodes provided by
+  # iroh, if not replaced) and any extra nodes defined by the user.
+  relaySettings = cfg.relayNodes // cfg.extraRelayNodes;
+
+  # The config file should be located in the data directory and is NOT created by default. All
+  # values are optional, including the file itself.
+  configFormat = pkgs.formats.toml { };
+  #configFile = configFormat.generate "${cfg.dataDir}/iroh.config.toml" relaySettings;
+  configFile = configFormat.generate "/var/lib/iroh/iroh.config.toml" relaySettings;
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:89]: this won't place a file under `/var/lib/iroh`

does `iroh` accept an immutable config file and directory? if so we should use the `IROH_CONFIG_DIR` environment variable to create a directory of configuration files...

maybe something like this: 

`environment.etc."iroh/iroh.config.toml".source = configFormat.generate "iroh.config.toml" relaySettings;`

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    #./other-module.nix
+  ];
+
+  # Option declarations for our module.
+  options = {
+    services.iroh = {
+
+      # By convention, modules have an "enable" option that determines whether they are active or
+      # not.
+      enable = mkEnableOption "Iroh is a protocol for syncing bytes. Bytes of any size, across any number of devices. Use iroh to create apps that scale different.";
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:114]: `mkEnableOption` prepends something like `"Whether to enable "` and appends `"."` to the text you pass to it

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+      package = mkPackageOption pkgs "iroh" { };
+
+      # NB: This is READ ONLY because there's no way to configure the listen port; iroh exposes as
+      # few configuration knobs as possible. The default port can be read from this read-only
+      # option, but if a different port is chosen by the service it has to be retrieved another way (?).
+      listenPort = mkOption {
+        description = "Address iroh listens on for connections from other nodes. If the port is taken iroh will choose a random port to listen on, so it can't be configured. This port is NOT automatically opened in the firewall.";
+        type = types.port;
+        readOnly = true;
+        default = listenPort;
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:125]: this value may not be correct so I'm not sure there is much value in having this option

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+    # Run iroh with a systemd "dynamic user":
+    # https://0pointer.net/blog/dynamic-users-with-systemd.html
+
+    # Define the systemd service.
+    systemd.services.iroh = {
+      wants = [ "network-online.target" ];
+      wantedBy = [ "default.target" ];
+
+      path = [ "/run/wrappers" pkgs.iroh ];
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:244]: why `"/run/wrappers"`?

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+          }
+          {
+            url = "https//euw1-1.derp.iroh.network.";
+            stunOnly = false;
+            stunPort = 3478;
+          }
+        ];
+      };
+
+      extraRelayNodes = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:181]: currently this is simply added to the other nodes - will it serve an additional purpose in the future?

File: nixos/modules/services/network-filesystems/iroh.nix [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+  # cfg is a conventional alias for the configuration of the current module
+  cfg = config.services.iroh;
+
+  # Default listening port.
+  listenPort = 11204;
+  # Default RPC port.
+  rpcPort = 1337;
+  # Default metrics port. With a value of -1 serving metrics is disabled.
+  metricsPort = -1;
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/iroh.nix:27]: no need to set these values here, you can just set them in `default` values below and reference them elsewhere as needed

**Review by aanderse — COMMENTED** [review]
thanks for the details!

i left a review despite this being a work in progress - i hope it is useful to you
let me know if you have any questions or i can help at all

**Comment by crimeminister** [comment]
> i left a review despite this being a work in progress - i hope it is useful to you let me know if you have any questions or i can help at all

Thanks for this! At this early stage in my experience with NixOS help is very welcome. I'll work on this in spare moments, but you may get sporadic questions as I learn what it is that I don't know :D

File: nixos/modules/services/network-filesystems/iroh.nix (hunk 2 above)
Comment (crimeminister) [reply nixos/modules/services/network-filesystems/iroh.nix:2]: I started with a copy/paste from elsewhere (mea culpa), seems as though I may have picked up some anachronisms. Is there a linter or repository of heuristics for good module development that you happen to know of?

Comment (aanderse) [reply nixos/modules/services/network-filesystems/iroh.nix:2]: @SuperSandro2000 is the best `nix` linter that i know of 😉

**Comment by crimeminister** [comment]
Closing as the underlying program this was using has been deprecated.
